### PR TITLE
health agent to monitor lotus

### PR DESCRIPTION
**Pull request by ognots**
watch if chain head changes in a given window of api polls
allows setting a threshold of how many times the chain head can remain
unchanged before failing health check
also can set interval for polling chain head
on failure, restarts systemd unit

### CODE REVIEW

File: cmd/lotus-health/main.go [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+	cid "github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log"
+	manet "github.com/multiformats/go-multiaddr-net"
+	"golang.org/x/xerrors"
+	"gopkg.in/urfave/cli.v2"
+)
+
+type CidWindow [][]cid.Cid
+
+var log = logging.Logger("lotus-seed")
```
Comment (magik6k) [inline cmd/lotus-health/main.go]:
```suggestion
var log = logging.Logger("lotus-health")
```

File: cmd/lotus-health/main.go [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+			Usage: "lotus repo path",
+		},
+		&cli.IntFlag{
+			Name:  "threshold",
+			Value: 3,
+			Usage: "number of times head remains unchanged before failing health check",
+		},
+		&cli.IntFlag{
+			Name:  "interval",
+			Value: 45,
```
Comment (magik6k) [inline cmd/lotus-health/main.go]:
```suggestion
			Value: build.BlockDelay,
```

File: cmd/lotus-health/main.go [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+		&cli.StringFlag{
+			Name:  "systemd-unit",
+			Value: "lotus-daemon.service",
+			Usage: "systemd unit name to restart on health check failure",
+		},
+	},
+	Action: func(c *cli.Context) error {
+		repo := c.String("repo")
+		threshold := c.Int("threshold")
+		interval := time.Duration(c.Int("interval"))
```
Comment (magik6k) [inline cmd/lotus-health/main.go]:
```suggestion
		interval := time.Duration(c.Int("interval")) * time.Second
```

File: cmd/lotus-health/main.go [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+		name := c.String("systemd-unit")
+
+		var headCheckWindow CidWindow
+		ctx := context.Background()
+
+		api, closer, err := GetFullNodeAPI(repo)
```
Comment (magik6k) [inline cmd/lotus-health/main.go]: Why not use `cli.GetFullNodeAPI(c)`?

File: cmd/lotus-health/main.go [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+		ch := make(chan CidWindow, 1)
+		aCh := make(chan interface{}, 1)
+
+		go func() {
+			for {
+				headCheckWindow, err = updateWindow(ctx, api, headCheckWindow, threshold, ch)
+				if err != nil {
+					log.Fatal(err)
+				}
+				time.Sleep(interval * time.Second)
```
Comment (magik6k) [inline cmd/lotus-health/main.go]:
```suggestion
				time.Sleep(interval)
```

File: cmd/lotus-health/main.go (hunk 4 above)
Comment (ognots) [reply cmd/lotus-health/main.go]: good idea, thanks!

File: Makefile [hunk 6]
```diff
@@ -149,6 +149,12 @@ stats:
 .PHONY: stats
 BINS+=stats
 
+health:
+	rm -f lotus-health
+	go build -o lotus-health ./cmd/lotus-health
```
Comment (magik6k) [inline Makefile]: Probably want
```suggestion
	go build -o lotus-health ./cmd/lotus-health
	go run github.com/GeertJohan/go.rice/rice append --exec lotus-health -i ./build
```

**Pull request merged**